akka-http-caching/src/main/scala/akka/http/caching/LfuCache.scala

```diff
@@ -111,7 +112,20 @@ private[caching] class LfuCache[K, V](val store: AsyncLoadingCache[K, V]) extend
 
   def apply(key: K, genValue: () ⇒ Future[V]): Future[V] = store.get(key, toJavaMappingFunction[K, V](genValue)).toScala
 
-  def getOrLoad(key: K, loadValue: K ⇒ Future[V]) = store.get(key, toJavaMappingFunction[K, V](loadValue)).toScala
+  def getOrLoad(key: K, loadValue: K ⇒ Future[V]): Future[V] =
+    store.get(key, toJavaMappingFunction[K, V](loadValue)).toScala
+
+  def put(key: K, mayBeValue: Future[V])(implicit ex: ExecutionContext): Future[Unit] = {
+    val cachedBefore = store.synchronous().asMap().containsKey(key)
+    val promise = Promise[Unit]
+    if (cachedBefore) {
+      mayBeValue.onComplete {
+        case Success(value)      ⇒ promise.completeWith(Future(store.synchronous().put(key, value)))
+        case failure: Failure[_] ⇒ promise.failure(failure.exception)
+      }
+    } else promise.completeWith(Future(store.put(key, toJava(mayBeValue).toCompletableFuture)))
```

Comment:
The logic of this 'else' branch seems OK, but perhaps it could be written as `store.put(key, toJava(mayBeValue).toCompletableFuture); Future.success(())`? That way we might not even need the `Promise`

What happens when `mayBeValue` fails here? Perhaps we should return `mayBeValue.map(_ => ())`?

Reply:
How about this implementation.
```scala
def put(key: K, mayBeValue: Future[V])(implicit ex: ExecutionContext): Future[Unit] = {
    val previouslyCacheValue = Option(store.getIfPresent(key))

    previouslyCacheValue match {
      case None ⇒ store.put(key, toJava(mayBeValue).toCompletableFuture)
        Future.successful(())
      case _ ⇒ mayBeValue.map(value ⇒ store.put(key, toJava(Future.successful(value)).toCompletableFuture))
    }
  }
```
Now in this implementation if the given future fails, the caller of this method can take appropiate action.